scri/bms_transformations.py

```diff
@@ -0,0 +1,829 @@
+            nvec = nvec_im
+            is_rotation = False
+        else:
+            spin_matrix_to_Lorentz(A)
+    else:
+        if is_rotation:
+            nvec = nvec_re
+        else:
+            nvec = nvec_im
+            
+    psi = 2 * np.linalg.norm(nvec)
+    if psi == 0:
+        nvec_hat = np.array([0,0,1])
+    else:
+        nvec_hat = nvec / np.linalg.norm(nvec)
+        
+    if is_rotation:
+        return np.quaternion(np.cos(psi / 2), *nvec_hat*np.sin(psi / 2)).components
+    else:
+        return np.tanh(psi) * nvec_hat
+    
+def spin_matrix_to_Lorentz(A, output_order=['rotation','boost']):
+    """Convert a spin matrix to a Lorentz transformation.
+
+    This uses SVD to decompose the spin matrix into a
+    spin matrix describing a Lorentz boost and a
+    spin matrix describing a Lorentz rotation.
+    
+    Parameters
+    ----------
+    A: 2 by 2 complex array
+        2 by 2 array corresponding to the input spin matrix.
+    output_order: list
+        Order in which rotation and boost should be applied.
+    """
+    if np.allclose(A, np.zeros_like(A)):
+        return Lorentz_transformation()
+    
+    u, s, vh = np.linalg.svd(A)
+    
+    try:
+        rotation_idx = output_order.index('rotation')
+    except:
+        rotation_idx = np.inf
+    try:
+        boost_idx = output_order.index('boost')
+    except:
+        boost_idx = np.inf
+        
+    if rotation_idx < boost_idx:
+        rotation_matrix = np.linalg.multi_dot([u, vh])
+        boost_matrix = np.linalg.multi_dot([u, np.diag(s), u.conj().T])
+    else:
+        rotation_matrix = np.linalg.multi_dot([u, vh])
+        boost_matrix = np.linalg.multi_dot([vh.conj().T, np.diag(s), vh])
+    rotation = pure_spin_matrix_to_Lorentz(rotation_matrix, is_rotation=True)
+    boost = pure_spin_matrix_to_Lorentz(boost_matrix, is_rotation=False)
+    return Lorentz_transformation(rotation=rotation, boost=boost, order=output_order)
+
+def compute_conformal_factor(Lorentz, n_theta, n_phi):
+    """Compute the conformal factor.
+
+    Note that this is the inverse of Eq. (4.6) of 10.1063/1.1705135.
+    """
+    thetaprm_phiprm = sf.theta_phi(n_theta, n_phi)
+
+    [[a, b], [c, d]] = Lorentz_to_spin_matrix(Lorentz)
+
+    conformal = np.zeros((n_theta, n_phi), dtype=complex)
+    for j in range(n_theta):
+        for k in range(n_phi):
+            thetaprm_j, phiprm_k = thetaprm_phiprm[j, k]
+            if thetaprm_j == 0:
+                z = np.inf
+                conformal[j, k] = 1/(np.abs(a)**2+np.abs(c)**2)
+            else:
+                z = np.exp(1j*phiprm_k)/np.tan(thetaprm_j/2)
+                conformal[j, k] = (1+np.abs(z)**2)/(np.abs(a*z+b)**2+np.abs(c*z+d)**2)
+            
+    return conformal
+
+def nCr(n, r):
+    """n choose r.
+    """
+    if n >= 0 and r >= 0 and n - r >= 0:
+        return np.math.factorial(n)/(np.math.factorial(r)*np.math.factorial(n-r))
+    else:
+        return 0
```

Comment:
Expanding out each factorial and then taking the product and ratio can lead to overflow or just loss of precision. `scipy.special.binom` and `scipy.special.comb` use methods to keep the numerator and denominator close to each other and minimize the number of multiplications, to minimize loss of precision.